platform/wab/src/wab/client/components/grouping/useTreeData.tsx

```diff
@@ -0,0 +1,285 @@
+import React from "react";
+import { Matcher } from "@/wab/client/components/view-common";
+import { mod } from "@/wab/shared/common";
+
+export type NodeKey = string;
+
+export type RenderElement<T> = React.FunctionComponent<RenderElementProps<T>>;
+
+type SelectDirection = 1 | -1;
+
+export interface RenderElementProps<T> {
+  value: T;
+  /** Tree state for the rendered component */
+  treeState: {
+    matcher: Matcher;
+    level: number;
+    canOpen: boolean;
+    isOpen: boolean;
+    isSelected?: boolean;
+    toggleExpand: () => void;
+  };
+}
+
+export interface LinearTreeNode<T> {
+  value: T;
+  key: NodeKey;
+  level: number;
+  parentKey?: NodeKey;
+  isOpen: boolean;
+  canOpen: boolean;
+}
+
+export interface TreeRowData<T> {
+  treeData: {
+    matcher: Matcher;
+    nodes: LinearTreeNode<T>[];
+    renderElement: RenderElement<T>;
+    selectedIndex?: number;
+    toggleExpand: (key: NodeKey) => void;
+  };
+}
+
+interface Props<T> {
+  nodes: T[];
+  query: string;
+  renderElement: RenderElement<T>;
+  getNodeKey: (node: T) => NodeKey;
+  getNodeChildren: (node: T) => T[];
+  getNodeSearchText: (node: T) => string;
+  getNodeHeight: (node: T) => number;
+  isNodeSelectable?: (node: T) => boolean;
+  defaultOpenKeys?: "all" | NodeKey[];
+}
+
+interface UseTreeData<T> {
+  nodeData: TreeRowData<T>;
+  nodeKey: (index: number, data: TreeRowData<T>) => string;
+  nodeHeights: number[];
+  selectedIndex?: number;
+  expandAll: () => void;
+  collapseAll: () => void;
+  selectNextRow: (direction: SelectDirection) => void;
+}
+
+export function useTreeData<T>({
+  nodes,
+  query = "",
+  renderElement,
+  getNodeKey,
+  getNodeChildren,
+  getNodeSearchText,
+  getNodeHeight,
+  isNodeSelectable,
+  defaultOpenKeys,
+}: Props<T>): UseTreeData<T> {
+  const matcher = React.useMemo(() => {
+    return new Matcher(query?.trim() ?? "");
+  }, [query]);
+  const [expandedNodes, setExpandedNodes] = React.useState<Set<NodeKey>>(
+    new Set(
+      defaultOpenKeys === "all"
+        ? getAllNodeKeys(nodes, getNodeKey, getNodeChildren)
+        : defaultOpenKeys ?? []
+    )
+  );
+  const toggleExpand = React.useCallback(
+    (key: NodeKey) => {
+      setExpandedNodes((set) => {
+        if (set.has(key)) {
+          set.delete(key);
+        } else {
+          set.add(key);
+        }
+        return new Set(set);
+      });
+    },
+    [setExpandedNodes]
+  );
+
+  const visibleNodes = React.useMemo<LinearTreeNode<T>[]>(
+    () =>
+      buildVisibleNodes(
+        nodes,
+        matcher,
+        expandedNodes,
+        getNodeKey,
+        getNodeChildren,
+        getNodeSearchText
+      ),
+    [
+      nodes,
+      matcher,
+      expandedNodes,
+      getNodeKey,
+      getNodeChildren,
+      getNodeSearchText,
+    ]
+  );
+
+  const [selectedIndex, setSelectedIndex] = React.useState<number>();
+
+  // Reset selectionIndex when visibleNodes change
```

Comment:
This seems messy, but I couldn't think of a better way to ensure there is a valid selection when the tree first opens.

Reply:
I explored this briefly, it looks like it will simplify some things but is a pretty substantial change. Would you rather I go for it here, or do it in parallel with other tasks (or leave it be for now)?